src/main/java/org/commcare/xml/bulk/BulkProcessingCaseXmlParser.java

```diff
@@ -0,0 +1,308 @@
+package org.commcare.xml.bulk;
+
+import org.commcare.cases.model.Case;
+import org.commcare.cases.model.CaseIndex;
+import org.javarosa.core.model.instance.TreeElement;
+import org.javarosa.core.model.utils.DateUtils;
+import org.javarosa.core.util.externalizable.SerializationLimitationException;
+import org.javarosa.xml.util.ActionableInvalidStructureException;
+import org.javarosa.xml.util.InvalidStructureException;
+import org.kxml2.io.KXmlParser;
+
+import java.util.Date;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+
+/**
+ * A parser which is capable of processing CaseXML transactions in a bulk format.
+ *
+ * This parser needs an implementation which can perform the "bulk" steps efficiently on the current
+ * platform.
+ *
+ * It should be a drop-in replacemenet for the CaseXmlParser in the ways that it is used
+ *
+ * Created by ctsims on 3/14/2017.
+ */
+public abstract class BulkProcessingCaseXmlParser extends BulkElementParser<Case> {
+
+    public BulkProcessingCaseXmlParser(KXmlParser parser) {
+        super(parser);
+    }
+
+    @Override
+    protected void requestModelReadsForElement(TreeElement bufferedTreeElement, Set<String> currentBulkReadSet) {
+        String caseId = bufferedTreeElement.getAttributeValue(null, "case_id");
+        currentBulkReadSet.add(caseId);
+    }
+
+    @Override
+    protected void preParseValidate() throws InvalidStructureException {
+        checkNode("case");
+    }
+
+    @Override
+    protected void processBufferedElement(TreeElement bufferedTreeElement, Map<String, Case> currentOperatingSet, LinkedHashMap<String, Case> writeLog) throws InvalidStructureException {
```

Comment:
Actually it looks like this is true for almost all of the methods in this class. I think this would feel a lot cleaner if all of those implementations were merged in some way. I don't like the idea that someone could go update `CaseXmlParser` down the line and not realize that this other place exists that should get the same changes. It's unfortunate that this class can't just inherit from `CaseXmlParser` as well, but some shared static methods don't seem terrible, since I think you should only have to pass the `parser` into each one.